expression/grouping_sets_test.go

```diff
@@ -0,0 +1,142 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"go.opencensus.io/stats/view"
+)
+
+func TestGroupSetsTargetOne(t *testing.T) {
+	defer view.Stop()
+	a := &Column{
+		UniqueID: 1,
+	}
+	b := &Column{
+		UniqueID: 2,
+	}
+	c := &Column{
+		UniqueID: 3,
+	}
+	d := &Column{
+		UniqueID: 4,
+	}
+	// non-merged group sets
+	//	1: group sets {<a,b>}, {<c>} when the normal agg(d), d can occur in either group of <a,b> or <c> after tuple split. (normal column are appended)
+	//  2: group sets {<a,b>}, {<c>} when the normal agg(c), c can only be found in group of <c>, since it will be filled with null in group of <a,b>
+	//  3: group sets {<a,b>}, {<c>} when the normal agg with multi col args, it will be a little difficult, which is banned in canUse3Stage4MultiDistinctAgg by now.
+	newGroupingSets := make(GroupingSets, 0, 3)
+	groupingExprs1 := []Expression{a, b}
+	groupingExprs2 := []Expression{c}
+	newGroupingSets = append(newGroupingSets, GroupingSet{groupingExprs1})
+	newGroupingSets = append(newGroupingSets, GroupingSet{groupingExprs2})
+
+	targetOne := []Expression{d}
+	offset := newGroupingSets.TargetOne(targetOne)
+	require.NotEqual(t, offset, -1)
+	require.Equal(t, offset, 0) // <a,b> and <c> both ok
+
+	targetOne = []Expression{c}
+	offset = newGroupingSets.TargetOne(targetOne)
+	require.NotEqual(t, offset, -1)
+	require.Equal(t, offset, 1) // only <c> is ok
+
+	targetOne = []Expression{b}
+	offset = newGroupingSets.TargetOne(targetOne)
+	require.NotEqual(t, offset, -1)
+	require.Equal(t, offset, 0) // only <a,b> is ok
+
+	targetOne = []Expression{a}
+	offset = newGroupingSets.TargetOne(targetOne)
+	require.NotEqual(t, offset, -1)
+	require.Equal(t, offset, 0) // only <a,b> is ok
+
+	targetOne = []Expression{b, c}
+	offset = newGroupingSets.TargetOne(targetOne)
+	require.Equal(t, offset, -1) // no valid one can be found.
+
+	// merged group sets
+	// merged group sets {<a>, <a,b>} when the normal agg(d),
+	// from prospect from data grouping, we need keep the widest grouping layout <a> when trying
+	// to shuffle data targeting for grouping layout {<a>, <a,b>}; Additional, since column b
+	// hasn't been pre-agg or something because current layout is <a>, we should keep all the
+	// complete rows until to the upper receiver, which means there is no agg/group operator before
+	// shuffler take place.
+}
+
+func TestGroupingSetsMergeUnitTest(t *testing.T) {
```

Comment:
This function is a strong algorithm logic, it is recommended to add more test examples.